Fix: MXP Sound/Music Tags not functioning with U Attribute

gcms opened the pull request:
Fixes issue #5876

<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->

#### Brief overview of PR changes/additions
- MXP tag parsing: Adds extra conditions for identifying closing tags with prefix, allowing parsing to continue when slashes are found inside an unquoted attribute value;  Test case for the example in the issue
- SOUND and MUSIC tag processing: extra check for detecting unnamed volume attribute
- Refactorings to remove duplicated code

#### Motivation for adding to Mudlet

#### Other info (issues closed, discussion etc)

#### Release post highlight
<!--
Use this space if you wish to write a short statement or example for inclusion
in the release post for the next release. 
-->

gcms committed:
Fixes issue #5876

- Adds extra conditions for identifying closing tags with prefix, allowing parsing to continue when slashes are found inside an unquoted attribute value

mpconley commented:
Thanks for checking this out!  Not working for me yet with these examples:
- `<sound FName="cow.wav" U=https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/>`
- `<sound FName="cow.wav" U="https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/">`

Note: I updated my test case on the game from my issue to use the quoted version as the standard hints that one is what should be used.

gcms committed:
Extra check if volume attribute in SOUND tag

- Volume attribute is expected to be the second attribute when attribute
names are not specified, however there was no check if it was an unnamed
attribute value

gcms committed:
Extra check of volume attribute in MUSIC tag

- Volume attribute is expected to be the second attribute when attribute
names are not specified, however there was no check if it was an unnamed
attribute value

gcms committed:
Refatoring to remove duplicated code

- Moves volume checks to TMediaData

gcms committed:
Refactoring to use simplified tag attributes acessors

SlySven commented on src/TMediaData.h:
```diff
@@ -54,7 +54,18 @@ class TMediaData
+        if (mediaVolume == TMediaData::MediaVolumePreload) {
+            // Support preloading
+            mMediaVolume = TMediaData::MediaVolumePreload;
+        } else if (mediaVolume > TMediaData::MediaVolumeMax) {
+            mMediaVolume = TMediaData::MediaVolumeMax;
+        } else if (mediaVolume < TMediaData::MediaVolumeMin) {
+            mMediaVolume = TMediaData::MediaVolumeMin;
+        } else {
+            mMediaVolume = mediaVolume;
+        }
```
Qt has a macro for this:
```suggestion
        } else {
            mMediaVolume =  qBound(TMediaData::MediaVolumeMin, mMediaVolume, TMediaData::MediaVolumeMax);
        }
```

SlySven commented on src/TMxpNodeBuilder.cpp:
```diff
@@ -185,7 +185,7 @@ bool TMxpNodeBuilder::acceptSequence(char ch, std::string& buffer)
             mHasSequence = true;
             return false;
-        } else if (ch == '/') {
+        } else if (ch == '/' && mCurrentTagAttrs.empty() && mCurrentAttrValue.empty()) {
```
```suggestion
        }
        if (ch == '/' && mCurrentTagAttrs.empty() && mCurrentAttrValue.empty()) {
```

SlySven reviewed (commented):
Normally I can't really comment on Zugg's ~~dark art~~ but the code fragments I can see don't *obviously* look wrong to me... I just have a couple of tweaks for them.

gcms committed:
Update src/TMxpNodeBuilder.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

gcms committed:
Update src/TMediaData.h

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

gcms committed:
Update TMediaData.h

gcms committed:
Cast of enums to int to use qBound

SlySven commented:
Sorry that my suggestion about `qbound(...)` didn't apply as cleanly as I'd have thought. Mind you, I *think* there is a way to tell the compiler in later C++ standards that the underlying (integer) type that is to be used is a particular flavour **+** (e.g. an `int32_t` rather than an `uint32_t`) by something at the start of the `enum` definition, though, contrary-wise if one is using it for bit-wise operations those are safer to carry out on unsigned integers rather than ones with signage.


**+** https://en.cppreference.com/w/cpp/language/enum - look for the `enum-base` bit!

vadi2 commented:
That is actually a really cool option

mpconley commented:
@vadi2 confirming this tests out and resolves the issue.  Tested non-MXP sound works too.  Thanks @gcms !

vadi2 commented:
We even ended up with less code! Thanks @gcms.
